From 599dc6a0ffdeb46ea9b4ca5759540e6f48cb31b1 Mon Sep 17 00:00:00 2001
From: Chih-Wei Huang <cwhuang@linux.org.tw>
Date: Wed, 15 Apr 2020 16:16:02 +0800
Subject: [PATCH] SurfaceFlinger: fix crashing if driver stalled

Due to some unknown reasons, the VSYNC may not be sent in time that
causes SurfaceFlinger to fake VSYNC like:

04-15 08:09:28.866  1235  1360 W SurfaceFlinger: Faking VSYNC due to driver stall

Therefore mPendingPeriod in DispSync may not be set to mPeriod which is
zero. It results in a dividing-by-zero error in SurfaceFlinger:

04-15 08:09:29.329  1395  1395 F DEBUG   : pid: 1235, tid: 1235, name: surfaceflinger  >>> /system/bin/surfaceflinger <<<
04-15 08:09:29.329  1395  1395 F DEBUG   : uid: 0
04-15 08:09:29.329  1395  1395 F DEBUG   : signal 8 (SIGFPE), code 1 (FPE_INTDIV), fault addr 0x7619c5aff344

To work around it, just use mPendingPeriod if mPeriod has not been set.

diff --git a/services/surfaceflinger/Scheduler/DispSync.cpp b/services/surfaceflinger/Scheduler/DispSync.cpp
index 94f363e7c..359edd3a6 100644
--- a/services/surfaceflinger/Scheduler/DispSync.cpp
+++ b/services/surfaceflinger/Scheduler/DispSync.cpp
@@ -661,7 +661,7 @@ nsecs_t DispSync::getPeriod() {
     if (mPendingPeriod && !mModelUpdated) {
         return mPendingPeriod;
     } else {
-        return mPeriod;
+        return mPeriod ? mPeriod : mPendingPeriod;
     }
 }
 
-- 
2.17.1

